mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/SolveMessageInconsistenciesService.java

```diff
@@ -445,11 +468,14 @@ Snapshot snapshot() {
             .switchIfEmpty(Mono.just(NO_INCONSISTENCY));
     }
 
-    private Flux<Task.Result> fixInconsistenciesInMessageId(Context context) {
+    private Flux<Task.Result> fixInconsistenciesInMessageId(Context context, RunningOptions runningOptions) {
         return messageIdDAO.retrieveAllMessages()
             .doOnNext(any -> context.incrementMessageIdEntries())
-            .concatMap(this::detectInconsistencyInMessageId)
-            .concatMap(inconsistency -> inconsistency.fix(context, messageIdToImapUidDAO, messageIdDAO));
+            .windowTimeout(runningOptions.getMessagesPerSecond(), Duration.ofSeconds(1))
+            .zipWith(Flux.interval(Duration.ofSeconds(1)))
+            .flatMap(Tuple2::getT1)
```

Comment:
Maybe we should write:

```
return throttle(messageIdDAO.retrieveAllMessages(), runningOptions)
    .doOnNext(any -> context.incrementMessageIdEntries())
    .flatMap(this::detectInconsistencyInMessageId)
    .flatMap(inconsistency -> inconsistency.fix(context, messageIdToImapUidDAO, messageIdDAO));
```

We should not mix reactor throttling technical stuff with our domain (fixing message inconsistencies) technical stuff.

I also wonder if throttling can't be exposed as part of ReactorUtils class. That would make updates easier when we come up with a better throttling algorithm...

Reply:
I personally think we should keep it as it is for now. After we are done with all the tasks then we will see what we can refactor later, because other tasks might have more parameters than just messagesPerSecond.